src/prof_sys.c

```diff
@@ -373,6 +373,7 @@ prof_dump_check_possible_error(prof_dump_arg_t *arg, bool err_cond,
 
 static int
 prof_dump_open_file_impl(const char *filename, int mode) {
+	unlink(filename);
```

Comment:
Can `unlink` fail, say if `filename` does not exist?

Reply:
I'm not too familiar with the `unlink` function, but we care about reentrancy for system calls. See how we handle other system calls in `include/jemalloc/internal/test_hooks.h`.

For failure handling, if `unlink` always return an error on new file name (when `prof_dump_file_overwrite` is off), then let's gate it.

Further, let's gate it at a higher level, in `prof_dump_open` instead of `prof_dump_open_impl`. The reason is unit testing. We want to test for how well we handle `unlink` failure. See how we define a function pointer for `prof_dump_open_impl` and how we test in `test/unit/prof_mdump.c`.

Reply:
> Can unlink fail, say if filename does not exist?
> For failure handling, if unlink always return an error on new file name (when prof_dump_file_overwrite is off), then let's gate it.

If `unlink` fails, it returns `-1` and sets `errno`. I'm intentionally not checking `errno` because:
 * If file doesn't exist (`ENOENT`), it's not a problem because we `creat` it on next line, and is likely intentional because `prof_dump_file_overwrite` isn't set
* If some other fs issue occurs (fs is readonly, io error, no kernel memory...) it's likely that the `creat` call immediately following will have the same issue, and its error states are handled by `prof_dump_check_possible_error` in `prof_dump_open`. If it's some error that will not affect `creat`, this is also OK - but suboptimal - as `creat` will truncate the overwrite file.

> Further, let's gate it at a higher level, in prof_dump_open instead of prof_dump_open_impl. The reason is unit testing. We want to test for how well we handle unlink failure.

So you're saying `unlink` should occur in `prof_dump_open` before `prof_dump_open_file` is called? Will look into the unit testing aspects

> I'm not too familiar with the unlink function, but we care about reentrancy for system calls

Looks like it's signal-safe (https://man7.org/linux/man-pages/man7/signal-safety.7.html)

Reply:
I believe your reasoning is correct, but let's not base our code on external assumptions.

What I was suggesting is basically something like the following (in addition to changing `include/jemalloc/internal/test_hooks.h`):

```c
static int
prof_dump_unlink_file_impl(const char *filename) {
        return unlink(filename);
}
prof_dump_unlink_file_t *JET_MUTABLE prof_dump_unlink_file =
    prof_dump_unlink_file_impl;

static int
prof_dump_open_file_impl(const char *filename, int mode) {
        return creat(filename, mode);
}
prof_dump_open_file_t *JET_MUTABLE prof_dump_open_file =
    prof_dump_open_file_impl;

static void
prof_dump_open(prof_dump_arg_t *arg, const char *filename) {
        if (prof_dump_file_overwrite) {
                int err = prof_dump_unlink_file(filename);
                prof_dump_check_possible_error(...);
        }
        arg->prof_dump_fd = prof_dump_open_file(filename, 0644);
        prof_dump_check_possible_error(arg, arg->prof_dump_fd == -1,
            "<jemalloc>: failed to open \"%s\"\n", filename);
}
```
And in the unit tests, you can point `prof_dump_unlink_file` to a mock function to verify that the error handling logic is working.

Thoughts?